contracts/data-storage/andromeda-counter/src/contract.rs

```diff
@@ -0,0 +1,272 @@
+        &deps.querier,
+        info,
+        BaseInstantiateMsg {
+            ado_type: CONTRACT_NAME.to_string(),
+            ado_version: CONTRACT_VERSION.to_string(),
+            kernel_address: msg.kernel_address,
+            owner: msg.owner,
+        },
+    )?;
+
+    RESTRICTION.save(deps.storage, &msg.restriction)?;
+
+    if let Some(initial_amount) = msg.initial_amount {
+        INITIAL_AMOUNT.save(deps.storage, &initial_amount)?;
+        CURRENT_AMOUNT.save(deps.storage, &initial_amount)?;
+    } else {
+        INITIAL_AMOUNT.save(deps.storage, &DEFAULT_INITIAL_AMOUNT)?;
+        CURRENT_AMOUNT.save(deps.storage, &DEFAULT_INITIAL_AMOUNT)?;
+    }
+
+    if let Some(increase_amount) = msg.increase_amount {
+        INCREASE_AMOUNT.save(deps.storage, &increase_amount)?;
+    } else {
+        INCREASE_AMOUNT.save(deps.storage, &DEFAULT_INCREASE_AMOUNT)?;
+    }
+
+    if let Some(decrease_amount) = msg.decrease_amount {
+        DECREASE_AMOUNT.save(deps.storage, &decrease_amount)?;
+    } else {
+        DECREASE_AMOUNT.save(deps.storage, &DEFAULT_DECREASE_AMOUNT)?;
+    }
+
+    Ok(resp)
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    let ctx = ExecuteContext::new(deps, info, env);
+    match msg {
+        ExecuteMsg::AMPReceive(pkt) => {
+            ADOContract::default().execute_amp_receive(ctx, pkt, handle_execute)
+        },
+        _ => handle_execute(ctx, msg),
+    }
+}
+
+fn handle_execute(ctx: ExecuteContext, msg: ExecuteMsg) -> Result<Response, ContractError> {
+
+    match msg.clone() {
+        ExecuteMsg::Increment {} => execute_increment(ctx),
+        ExecuteMsg::Decrement {} => execute_decrement(ctx),
+        ExecuteMsg::Reset {} => execute_reset(ctx),
+        ExecuteMsg::UpdateRestriction { restriction } => execute_update_restriction(ctx, restriction),
+        ExecuteMsg::SetIncreaseAmount { increase_amount } => execute_set_increase_amount(ctx, increase_amount),
+        ExecuteMsg::SetDecreaseAmount { decrease_amount } => execute_set_decrease_amount(ctx, decrease_amount),
+        _ => ADOContract::default().execute(ctx, msg),
+
+    }
+}
+
+pub fn execute_increment(ctx: ExecuteContext) -> Result<Response, ContractError> {
+    nonpayable(&ctx.info)?;
+
+    let sender = ctx.info.sender.clone();
+    ensure!(
+        has_permission(ctx.deps.storage, &sender)?,
+        ContractError::Unauthorized {}
+    );
+    let current_amount = CURRENT_AMOUNT.load(ctx.deps.storage)?;
+    let increase_amount = INCREASE_AMOUNT.load(ctx.deps.storage)?;
+    CURRENT_AMOUNT.save(ctx.deps.storage, &(current_amount + increase_amount))?;
+
+    Ok(
+        Response::new()
+        .add_attribute("method", "execute_increment")
+        .add_attribute("sender", sender)
+        .add_attribute("current_amount", &CURRENT_AMOUNT.load(ctx.deps.storage)?.to_string())
```

Comment:
Would be more efficient to store the sum of `current_amount + increase_amount` in a variable to avoid loading from storage